NorthstarDLL/mods/autodownload/moddownloader.cpp

```diff
@@ -0,0 +1,552 @@
+		std::string cla = std::string(clachar);
+		if (strncmp(cla.substr(iFlagStringLength, 1).c_str(), "\"", 1))
+		{
+			int space = cla.find(" ");
+			url = cla.substr(iFlagStringLength, space - iFlagStringLength);
+		}
+		else
+		{
+			std::string quote = "\"";
+			int quote1 = cla.find(quote);
+			int quote2 = (cla.substr(quote1 + 1)).find(quote);
+			url = cla.substr(quote1 + 1, quote2);
+		}
+		spdlog::info("Found custom verified mods URL in command line argument: {}", url);
+		modsListUrl = strdup(url.c_str());
+	}
+	else
+	{
+		spdlog::info("Custom verified mods URL not found in command line arguments, using default URL.");
+		modsListUrl = strdup(DEFAULT_MODS_LIST_URL);
+	}
+}
+
+size_t WriteToString(void* ptr, size_t size, size_t count, void* stream)
+{
+	((std::string*)stream)->append((char*)ptr, 0, size * count);
+	return size * count;
+}
+
+void ModDownloader::FetchModsListFromAPI()
+{
+	std::thread requestThread(
+		[this]()
+		{
+			CURLcode result;
+			CURL* easyhandle;
+			rapidjson::Document verifiedModsJson;
+			std::string url = modsListUrl;
+
+			curl_global_init(CURL_GLOBAL_ALL);
+			easyhandle = curl_easy_init();
+			std::string readBuffer;
+
+			// Fetching mods list from GitHub repository
+			curl_easy_setopt(easyhandle, CURLOPT_CUSTOMREQUEST, "GET");
+			curl_easy_setopt(easyhandle, CURLOPT_TIMEOUT, 30L);
+			curl_easy_setopt(easyhandle, CURLOPT_URL, url.c_str());
+			curl_easy_setopt(easyhandle, CURLOPT_FAILONERROR, 1L);
+			curl_easy_setopt(easyhandle, CURLOPT_WRITEDATA, &readBuffer);
+			curl_easy_setopt(easyhandle, CURLOPT_WRITEFUNCTION, WriteToString);
+			result = curl_easy_perform(easyhandle);
+
+			if (result == CURLcode::CURLE_OK)
+			{
+				spdlog::info("Mods list successfully fetched.");
+			}
+			else
+			{
+				spdlog::error("Fetching mods list failed: {}", curl_easy_strerror(result));
+				goto REQUEST_END_CLEANUP;
+			}
+
+			// Load mods list into local state
+			spdlog::info("Loading mods configuration...");
+			verifiedModsJson.Parse(readBuffer);
+			for (auto i = verifiedModsJson.MemberBegin(); i != verifiedModsJson.MemberEnd(); ++i)
+			{
+				std::string name = i->name.GetString();
+				std::string dependency = i->value["DependencyPrefix"].GetString();
+
+				std::unordered_map<std::string, VerifiedModVersion> modVersions;
+				rapidjson::Value& versions = i->value["Versions"];
+				assert(versions.IsArray());
+				for (auto& attribute : versions.GetArray())
+				{
+					assert(attribute.IsObject());
+					std::string version = attribute["Version"].GetString();
+					std::string checksum = attribute["Checksum"].GetString();
+					modVersions.insert({version, {.checksum = checksum}});
+				}
+
+				VerifiedModDetails modConfig = {.dependencyPrefix = dependency, .versions = modVersions};
+				verifiedMods.insert({name, modConfig});
+				spdlog::info("==> Loaded configuration for mod \"" + name + "\"");
+			}
+
+			spdlog::info("Done loading verified mods list.");
+
+		REQUEST_END_CLEANUP:
```

Comment:
using a function is a bit annoying since you would have to pass the relevant data structures over and they may not all want to do the same thing.

Though it is pointless in `FetchModSync`

For example:
- `FetchModsListFromAPI` only wants to clean up curl
- `IsModLegit` wants to clean up the hash handle
- `ExtractMod` wants to remove leftover data

You could copy all the relevant logic to each error case, but that would often result in a lot of redundant copies and a different control flow.